File: traffic_cypher_in_C/src_c/entropy_extractor.c
#include "entropy_extractor.h"

#include <stdlib.h>
#include <string.h>
#include <math.h>
#include <openssl/evp.h>

static void sha256_hash(const uint8_t *data, size_t len, uint8_t out[32]) {
    EVP_MD_CTX *ctx = EVP_MD_CTX_new();
    EVP_DigestInit_ex(ctx, EVP_sha256(), NULL);
    EVP_DigestUpdate(ctx, data, len);
    unsigned int md_len;
    EVP_DigestFinal_ex(ctx, out, &md_len);
    EVP_MD_CTX_free(ctx);
}

extracted_entropy_t extract_entropy(const uint8_t *current_data, size_t current_len,
                                    const uint8_t *previous_data, size_t previous_len,
                                    uint32_t width, uint32_t height) {
    extracted_entropy_t result;
    memset(&result, 0, sizeof(result));

    /* Max size: 32 (full hash) + 32 (delta hash) + 64*32 (spatial) = 2112 */
    size_t max_entropy = 32 + 32 + 64 * 32;
    result.entropy_bytes = (uint8_t *)malloc(max_entropy);
    result.entropy_len = 0;

    /* 1. Full-frame hash */
    uint8_t full_hash[32];
    sha256_hash(current_data, current_len, full_hash);
    memcpy(result.entropy_bytes + result.entropy_len, full_hash, 32);
    result.entropy_len += 32;

    /* 2. Inter-frame delta hash + metrics */
    if (previous_data && previous_len > 0) {
        size_t min_len = current_len < previous_len ? current_len : previous_len;
        uint8_t *delta = (uint8_t *)malloc(min_len);
        uint64_t changed_pixels = 0;
        uint64_t total_delta = 0;

        for (size_t i = 0; i < min_len; i++) {
            delta[i] = current_data[i] ^ previous_data[i];
            if (delta[i] != 0) changed_pixels++;
            int16_t diff = (int16_t)current_data[i] - (int16_t)previous_data[i];
            total_delta += (uint64_t)(diff < 0 ? -diff : diff);
        }

        uint8_t delta_hash[32];
        sha256_hash(delta, min_len, delta_hash);
        memcpy(result.entropy_bytes + result.entropy_len, delta_hash, 32);
        result.entropy_len += 32;

        result.metrics.has_metrics = 1;
        result.metrics.changed_pixel_ratio = (double)changed_pixels / (double)min_len;
        result.metrics.mean_pixel_delta = (double)total_delta / (double)min_len;

        free(delta);
    }

    /* 3. Spatial block hashes (8x8 grid) */
    uint32_t grid_cols = 8, grid_rows = 8;
    uint32_t block_w = width / grid_cols;
    uint32_t block_h = height / grid_rows;

    if (block_w > 0 && block_h > 0) {
        for (uint32_t row = 0; row < grid_rows; row++) {
            for (uint32_t col = 0; col < grid_cols; col++) {
                EVP_MD_CTX *ctx = EVP_MD_CTX_new();
                EVP_DigestInit_ex(ctx, EVP_sha256(), NULL);

                for (uint32_t y = row * block_h; y < (row + 1) * block_h; y++) {
                    size_t line_start = (size_t)(y * width * 3 + col * block_w * 3);
                    size_t line_end = line_start + (size_t)(block_w * 3);
                    if (line_end <= current_len) {
                        EVP_DigestUpdate(ctx, current_data + line_start, block_w * 3);
                    }
                }

                uint8_t block_hash[32];
                unsigned int md_len;
                EVP_DigestFinal_ex(ctx, block_hash, &md_len);
                EVP_MD_CTX_free(ctx);

                memcpy(result.entropy_bytes + result.entropy_len, block_hash, 32);
                result.entropy_len += 32;
            }
        }
    }

    return result;
}

File: traffic_cypher_in_Rust/src/entropy_extractor.rs
use sha2::{Digest, Sha256};

/// Metrics about frame-to-frame visual variation.
#[derive(Debug, Clone)]
pub struct EntropyMetrics {
    /// Fraction of pixels that changed compared to previous frame (0.0 to 1.0)
    pub changed_pixel_ratio: f64,
    /// Mean absolute pixel delta across all channels
    pub mean_pixel_delta: f64,
}

/// Result of extracting entropy from a single frame.
pub struct ExtractedEntropy {
    /// Concatenated SHA-256 hashes forming the raw entropy
    pub entropy_bytes: Vec<u8>,
    /// Optional metrics (only if a previous frame was provided)
    pub metrics: Option<EntropyMetrics>,
}

/// Extract entropy from a frame's raw pixel data.
///
/// 1. SHA-256 of full pixel data → base entropy (32 bytes)
/// 2. If previous frame given: XOR delta → SHA-256 → inter-frame entropy (32 bytes)
/// 3. 8×8 spatial grid of block hashes → 64 × 32 = 2048 bytes of spatial entropy
///
/// Returns concatenated hashes as raw entropy.
pub fn extract_entropy(
    current_data: &[u8],
    previous_data: Option<&[u8]>,
    width: u32,
    height: u32,
) -> ExtractedEntropy {
    let mut entropy_bytes = Vec::with_capacity(32 + 32 + 64 * 32);

    // 1. Full-frame hash
    let full_hash = sha256(current_data);
    entropy_bytes.extend_from_slice(&full_hash);

    // 2. Inter-frame delta hash + metrics
    let metrics = if let Some(prev) = previous_data {
        let min_len = current_data.len().min(prev.len());
        let mut delta = vec![0u8; min_len];
        let mut changed_pixels = 0u64;
        let mut total_delta = 0u64;

        for i in 0..min_len {
            delta[i] = current_data[i] ^ prev[i];
            if delta[i] != 0 {
                changed_pixels += 1;
            }
            total_delta += (current_data[i] as i16 - prev[i] as i16).unsigned_abs() as u64;
        }

        let delta_hash = sha256(&delta);
        entropy_bytes.extend_from_slice(&delta_hash);

        let pixel_count = min_len / 3; // RGB
        Some(EntropyMetrics {
            changed_pixel_ratio: changed_pixels as f64 / min_len as f64,
            mean_pixel_delta: total_delta as f64 / min_len as f64,
        })
    } else {
        None
    };

    // 3. Spatial block hashes (8x8 grid)
    let grid_cols = 8u32;
    let grid_rows = 8u32;
    let block_w = width / grid_cols;
    let block_h = height / grid_rows;

    if block_w > 0 && block_h > 0 {
        for row in 0..grid_rows {
            for col in 0..grid_cols {
                let mut hasher = Sha256::new();
                for y in (row * block_h)..((row + 1) * block_h) {
                    let line_start = (y * width * 3 + col * block_w * 3) as usize;
                    let line_end = line_start + (block_w * 3) as usize;
                    if line_end <= current_data.len() {
                        hasher.update(&current_data[line_start..line_end]);
                    }
                }
                let block_hash = hasher.finalize();
                entropy_bytes.extend_from_slice(&block_hash);
            }
        }
    }

    ExtractedEntropy {
        entropy_bytes,
        metrics,
    }
}

fn sha256(data: &[u8]) -> [u8; 32] {
    let mut hasher = Sha256::new();
    hasher.update(data);
    let result = hasher.finalize();
    result.into()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_different_data_different_entropy() {
        let data_a = vec![10u8; 320 * 240 * 3];
        let data_b = vec![200u8; 320 * 240 * 3];

        let ea = extract_entropy(&data_a, None, 320, 240);
        let eb = extract_entropy(&data_b, None, 320, 240);

        assert_ne!(ea.entropy_bytes, eb.entropy_bytes);
    }

    #[test]
    fn test_identical_data_same_entropy() {
        let data = vec![42u8; 320 * 240 * 3];
        let ea = extract_entropy(&data, None, 320, 240);
        let eb = extract_entropy(&data, None, 320, 240);

        assert_eq!(ea.entropy_bytes, eb.entropy_bytes);
    }

    #[test]
    fn test_inter_frame_delta_metrics() {
        let data_a = vec![100u8; 320 * 240 * 3];
        let data_b = vec![150u8; 320 * 240 * 3];

        let result = extract_entropy(&data_b, Some(&data_a), 320, 240);
        let metrics = result.metrics.unwrap();

        assert!(metrics.changed_pixel_ratio > 0.99); // all pixels changed
        assert!((metrics.mean_pixel_delta - 50.0).abs() < 0.01);
    }
}
